src/python/rs_bam_handling.py

```diff
@@ -0,0 +1,90 @@
+    # final input will be of form (output_) grouping.bam 
+    # if any of tags are not present in the groupings they will be left as temp files
+    # will place results in directory named output
+    if output:
+        directoryName = output+ "_files"
+    else:
+        directoryName = bamFile[0:-4] + "_tag_files"
+    if not os.path.exists(directoryName):
+        os.makedirs(directoryName)
+    dirPath = str(pathlib.Path().resolve()) + "/" + directoryName + "/"
+    if mapping:
+        mappingTagsMissing = []
+        pysam.samtools.split(bamFile, "-d", sortTarget, "-@", str(threadNumber), "-u", dirPath + "untagged.bam", "--output-fmt", "BAM", "-f", dirPath + ("'" + "%!.bam'"), catch_stdout=False)
+        tempFiles = []
+        if type(mapping) == dict:
+            if output:
+                for cellType in mapping.keys():
+                    targetList = [dirPath + ("'" + "_" + tag + ".bam'") for tag in mapping[cellType]]
+                    tempFiles.extend(targetList)
+                    tempFiles.extend(targetList)
+                    presentList = [tag for tag in targetList if Path(tag).is_file()]
+                    mappingTagsMissing.extend(list(set(targetList) - set(presentList)))
+                    if len(presentList) > 0:
+                        pysam.merge("-f", "-@", str(threadNumber), "-o", dirPath + output + cellType + ".bam", *targetList)
+            else:
+                for cellType in mapping.keys():
+                    targetList = [dirPath + ("'"+ tag + ".bam'")  for tag in mapping[cellType]]
+                    tempFiles.extend(targetList)
+                    tempFiles.extend(targetList)
+                    presentList = [tag for tag in targetList if Path(tag).is_file()]
+                    mappingTagsMissing.extend(list(set(targetList)- set(presentList)))
+                    if len(presentList) > 0:
+                        pysam.merge("-f", "-@", str(threadNumber),  "-o", dirPath + cellType + ".bam", *targetList)
+        else:
+            mappingDict = {}
+            mappingDf = pd.read_csv(mapping, sep = delim)
+            clusters =  mappingDf[mappingColumns[0]].unique()
+            for cluster in clusters:
+                mappingDict[cluster] = mappingDf.loc[mappingDf[mappingColumns[0]] == cluster][mappingColumns[1]].unique()
+
+            if output:
+                for cellType in mappingDict.keys():
+                    targetList = [dirPath + ("'" + tag + ".bam'") for tag in mappingDict[cellType]]
+                    tempFiles.extend(targetList)
+                    presentList = [tag for tag in targetList if Path(tag).is_file()]
+                    mappingTagsMissing.extend(list(set(targetList)- set(presentList)))
+                    if len(presentList) > 0:
+                        pysam.merge("-f", "-@", str(threadNumber), "-o", dirPath + output + cellType + ".bam", *targetList)
```

Comment:
Not quite sure what the value is of repeating the output prefix in both the directory name and filename